### ISSUE TITLE
DataFlowAnalysis: method invocation on struct is reported as write operation

### ISSUE THREAD

**Issue by bkoelman** [dotnet/roslyn#14886]
**Version Used**: 
NuGet Microsoft.CodeAnalysis, v2.0.0-beta5

**Steps to Reproduce**:
Consider the next program:
```csharp
using System;
using System.Linq;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;

namespace RoslynDataFlowTester
{
    class Program
    {
        static void Main(string[] args)
        {
            string code = @"
                public struct S
                {
                    public void M()
                    {
                    }
                }

                public class C
                {
                    public void N(S s)
                    {
                        s.M();
                    }
                }
            ";

            var document = new AdhocWorkspace()
                .AddProject("Test", LanguageNames.CSharp)
                .WithCompilationOptions(new CSharpCompilationOptions(outputKind: OutputKind.DynamicallyLinkedLibrary))
                .AddMetadataReferences(new[] {
                    MetadataReference.CreateFromFile(typeof (object).Assembly.Location),
                })
                .AddDocument("Test.cs", code);

            Compilation compilation = document.Project.GetCompilationAsync().Result;

            var type = compilation.GetTypeByMetadataName("C");
            var method = (IMethodSymbol)type.GetMembers("N").First();
            var parameter = method.Parameters[0];

            var syntaxTree = type.DeclaringSyntaxReferences[0].GetSyntax().SyntaxTree;
            var model = compilation.GetSemanticModel(syntaxTree);
            SyntaxNode body = ((MethodDeclarationSyntax)method.DeclaringSyntaxReferences[0].GetSyntax()).Body;

            DataFlowAnalysis dataFlowAnalysis = model.AnalyzeDataFlow(body);
            if (dataFlowAnalysis.Succeeded)
            {
                if (dataFlowAnalysis.WrittenInside.Contains(parameter))
                {
                    Console.WriteLine("Write to 'p' detected.");
                }
                else
                {
                    Console.WriteLine("No writes to 'p' detected.");
                }
            }
        }
    }
}
```

**Expected Behavior**:
"No writes to 'p' detected." is written to the console.

**Actual Behavior**:
"Write to 'p' detected." is written to the console.

Changing the type of `S` to a `class` results in the expected behavior. I'm observing the same issue in NuGet Microsoft.CodeAnalysis, v1.3.2 and VS 2015 Update 3. 

If this is not a bug, can anyone please explain why a write is reported in the `struct` case?

**Comment by gafter** [comment]
This is correct behavior. The implicit `this` parameter of a method in a struct is of type `ref S` and can therefore write the struct itself. For a class, however, the `this` parameter is of type `C` and therefore a method cannot change the reference held by the caller.

**Comment by bkoelman** [comment]
I noticed that integral types such as `int`, `bool` are not reported to be written to. Are they special-cased? 

I've browsed through the inner workings, but could not find anything to indicate that.

**Comment by CyrusNajmabadi** [comment]
> I noticed that integral types such as int, bool are not reported to be written to. Are they special-cased?

Yes.  These types are assumed to have well behaved implementation in the runtime such that method calls on them do not mutate state.  This avoids unneeded copying when doing things like operating on readonly instances of them.

Such assumptions are not safe to make around general user structs.
